File: core/grid.py
"""Core domain model: Grid

Contains the Grid class representing the puzzle grid.
"""
from core.position import Position
from core.cell import Cell
from core.adjacency import Adjacency

class Grid:
    """A grid is a basic board container. Fields:
    rows: int, cols: int, cells: List[List[Cell]], adjacency: Adjacency"""
    def __init__(self, rows: int, cols: int, allow_diagonal: bool = True):
        self.rows = rows
        self.cols = cols
        self.cells = [[Cell(Position(r, c)) for c in range(cols)] for r in range(rows)]
        self.adjacency = Adjacency(allow_diagonal=allow_diagonal)
    def get_cell(self, pos: Position) -> Cell:
        """Returns the cell at the given position."""
        if 0 <= pos.row < self.rows and 0 <= pos.col < self.cols:
            return self.cells[pos.row][pos.col]
        else:
            raise IndexError("Position out of grid bounds.")
    def set_cell_value(self, pos: Position, value: int):
        """Sets the value of the cell at the given position."""
        cell = self.get_cell(pos)
        if not cell.given and not cell.blocked:
            cell.value = value
        else:
            raise ValueError("Cannot set value of a blocked cell.")
    def clear_cell(self, pos: Position):
        """Clears the value of the cell at the given position, but leaves given alone."""
        cell = self.get_cell(pos)
        if not cell.given and not cell.blocked:
            cell.value = None
        else:
            raise ValueError("Cannot clear a given cell.")
    def iter_cells(self):
        """Yields every Cell in the grid, 
        in row-major order (top to bottom, left to right)."""
        for row in self.cells:
            for cell in row:
                yield cell
    def empty_positions(self):
        """Returns a list (or iterator) of all positions that have no value assigned (i.e., not given, not filled)."""
        for cell in self.iter_cells():
            if cell.is_empty():
                yield cell.pos
    def filled_positions(self):
        """Returns a list (or iterator) of all positions that currently have a value (either givens or solver/generator placements)."""
        for cell in self.iter_cells():
            if cell.is_filled():
                yield cell.pos
    def print_grid(self):
        """Prints a simple text representation of the grid to the console."""
        for row in self.cells:
            row_str = ""
            for cell in row:
                if cell.blocked:
                    row_str += " X "
                elif cell.value is not None:
                    row_str += f" {cell.value} "
                else:
                    row_str += " . "
            print(row_str)

File: frontend/src/domain/grid.ts
/**
 * Grid domain model for Hidato puzzle state.
 */

export interface Cell {
  row: number;
  col: number;
  given: boolean;
  value: number | null;
  candidates: number[];
}

export interface Grid {
  size: number;
  cells: Cell[][];
}

/**
 * Create an empty grid of given size.
 */
export function createEmptyGrid(size: number): Grid {
  const cells: Cell[][] = [];
  for (let row = 0; row < size; row++) {
    cells[row] = [];
    for (let col = 0; col < size; col++) {
      cells[row][col] = {
        row,
        col,
        given: false,
        value: null,
        candidates: [],
      };
    }
  }
  return { size, cells };
}

/**
 * Get a cell from the grid safely.
 */
export function getCell(grid: Grid, row: number, col: number): Cell | null {
  if (row < 0 || row >= grid.size || col < 0 || col >= grid.size) {
    return null;
  }
  return grid.cells[row][col];
}

/**
 * Check if a position is within grid bounds.
 */
export function isInBounds(grid: Grid, row: number, col: number): boolean {
  return row >= 0 && row < grid.size && col >= 0 && col < grid.size;
}

/**
 * Get all 8-neighbor positions for a cell.
 */
export function getNeighbors(
  grid: Grid,
  row: number,
  col: number
): Array<{ row: number; col: number }> {
  const neighbors: Array<{ row: number; col: number }> = [];
  for (let dr = -1; dr <= 1; dr++) {
    for (let dc = -1; dc <= 1; dc++) {
      if (dr === 0 && dc === 0) continue;
      const nr = row + dr;
      const nc = col + dc;
      if (isInBounds(grid, nr, nc)) {
        neighbors.push({ row: nr, col: nc });
      }
    }
  }
  return neighbors;
}
